lib/archethic/contracts/interpreter/scope.ex

```diff
@@ -19,27 +21,109 @@ defmodule Archethic.Contracts.Interpreter.Scope do
   @doc """
   Create a new nested scope
   """
-  @spec create(list(String.t())) :: :ok
-  def create(scope_hierarchy) do
+  @spec create() :: :ok
+  def create() do
+    current_context = get_current_context()
+    current_scope_hierarchy = get_context_scope_hierarchy(current_context)
+    ref = new_ref()
+
+    Process.put(
+      :scope,
+      put_in(Process.get(:scope), [current_context] ++ current_scope_hierarchy ++ [ref], %{})
+    )
+
+    Process.put(
+      :scope,
+      update_in(Process.get(:scope), [current_context, "scope_hierarchy"], &(&1 ++ [ref]))
+    )
+
+    :ok
+  end
+
+  @doc """
+  Create new context
+  """
+  @spec create_context() :: :ok
+  def create_context() do
+    context_ref = new_ref()
+
+    new_context = %{
+      "scope_hierarchy" => []
+    }
+
+    # add context to scope and update context list
+    new_scope =
+      Process.get(:scope)
+      |> Map.put(context_ref, new_context)
+      |> Map.update!("context_list", &[context_ref | &1])
+
+    Process.put(:scope, new_scope)
+
+    :ok
+  end
+
+  @doc """
+  Leave a scope by removing it from current context's scope hierarchy and deleting its content
+  """
+  @spec leave_scope() :: :ok
+  def leave_scope() do
+    current_context = get_current_context()
+
     Process.put(
       :scope,
-      put_in(Process.get(:scope), scope_hierarchy, %{})
+      update_in(
+        Process.get(:scope),
+        [current_context, "scope_hierarchy"],
+        &List.delete_at(&1, -1)
+      )
     )
 
     :ok
   end
 
+  @doc """
+  Leave a context by removing it from scope and context_list
+  """
+  @spec leave_context() :: :ok
+  def leave_context() do
+    current_context = get_current_context()
+
+    new_scope =
+      Process.get(:scope)
+      |> Map.delete(current_context)
+      |> Map.update!("context_list", &List.delete_at(&1, 0))
```

Comment:
Here you could replace `&List.delete_at(&1, 0)` by pattern matching like `fn [_first | reset] -> rest end`